### ISSUE TITLE
The source of `@woocommerce/settings` processed by DEWP is confusing and not importable

### ISSUE THREAD

**Issue by tomalec** [woocommerce/woocommerce#35603]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

*This is a clone of [https://github.com/woocommerce/woocommerce-admin/issues/7810](<https://github.com/woocommerce/woocommerce-admin/issues/7810>)* as that was closed and archived, without serving a solution or linking to an issue which would somewhat describe potential improvement in that area

As a plugin developer, I tried to use Dependency Extraction Webpack Plugin to use a package to obtain the WooCommerce settings. ([woocommerce/pinterest-for-woocommerce#224](<https://github.com/woocommerce/pinterest-for-woocommerce/pull/224>), [woocommerce/google-listings-and-ads#1050](<https://github.com/woocommerce/google-listings-and-ads/issues/1050>))

I checked that `wc.wcSettings` delivers the data I need. So I checked the [README of DEWP](<https://github.com/woocommerce/woocommerce-admin/blob/main/packages/dependency-extraction-webpack-plugin/README.md?rgh-link-date=2021-10-18T10%3A43%3A28Z>) to check which plugin I should actually import in my local `package.json`, and which package I should inspect to check its API and behavior. README states:

| Request | Global | Script handle |
| -- | -- | -- |
| @woocommerce/settings | wc\['wcSettings'\] | wc-settings |

But the thing is, it's not the [`@woocommerce/settings`](<https://www.npmjs.com/package/@woocommerce/settings>). It is `wp-content/plugins/woocommerce/packages/woocommerce-blocks/build/wc-settings.js`

Which AFAIK is not published anywhere, therefore, could not even be my plugins dependency.

### Expected behavior

The README points to the exact package that would be externalized.

### Actual behavior

The local `@woocommerce/settings` package is different than externalized.

Even though, I can localize the source of that externalized import by reading the `woocommerce/woocommerce` source code, I still cannot import it in my local tests.

### Steps to reproduce

1. Install and setup Woo DEWP on your plugin repo.
2. Add `import foo from @woocommerce/settings`
3. Install locally `npm install --save @woocommerce/settings`
4. Run unit tests

### WordPress Environment

n/a

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](<https://woocommerce.com/storefront/>).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by samueljseay** [comment]
@tomalec thanks for raising this. I just want to make sure I understand. 

So it sounds like there are 2 issues:

1.  the package you wanted to import is not actually published anywhere, so you have to load it via DEWP. Yes it makes sense this would still work as long as you have WooCommerce core loaded, since DEWP will load it from the asset manifest

2. The README suggests there is a package available and published that is not?

Is that correct?

@psealock I'm wondering if maybe you can give some advice, are there some packages on npm that haven't been properly re-released since going to the monorepo? For example https://www.npmjs.com/package/@woocommerce/settings points back to the deprecated wc admin repo, is that just something we need to fix for releasing new versions of npm packages?

**Comment by psealock** [comment]
`import { foo } from @woocommerce/settings` will be replaced by DEWP to be `const foo = wc.wcSettings.foo`. You're right the package doesn't actually exist. In theory, even though you `npm install --save` a package, its not actually getting used, but instead read off the global scope.

It sounds like the problem you are experiencing is related to testing. Indeed, without an actual package a problem will occur. Testing using data from the global `wcSettings` can be problematic because that data can be derived from multiple places or plugins which may not be present in your testing environment. WC-Admin gets around this by doing some variation of this. Is this helpful at all?

https://github.com/woocommerce/woocommerce/blob/1550806efc513643d7a4ae7fed0bd723e2c62929/plugins/woocommerce-admin/client/navigation/components/header/test/index.js#L12-L23

**Comment by tomalec** [comment]
> So it sounds like there are 2 issues: […]

The problem is quite composite, and it's hard for me to even list exact points, as to me (wearing the extension developer hat) the entire way the "settings package" (I don't even know how to precisely reference it) is delivered & documented is more confusing than beneficial.

1. If I figure out that I need to use `wc.wcSettings` as per DEWP readme I start using `import { foo } from @woocommerce/settings`. Then to check the details of `@woocommerce/settings` I look up in npm. As I do for all the other packages, and all the other imports with the bare specifier.
   However, this gives me irrelevant, misleading information. So maybe, from the `woocommerce/woocommerce` perspective, the package name should be different, not to collide with an existing package?
2. Now, once I realized that `@woocommerce/settings` is not [`@woocommerce/settings`](https://www.npmjs.com/package/@woocommerce/settings), where do I look up for docs for it?
   I'd like to find that easier and shorter than after reading and understanding the DEWP source code, and browsing the entire `woocommerce/woocommerce` repo.

Also, without having any `@woocommerce/setting` installed :

3. `@woocommerce/eslint-plugin/recommended` throws errors.
4. I don't get any IntelliSense support.
5. Tests are failing 

To mitigate failing tests we mocked it with jest config:
```
		'@woocommerce/settings':
			'<rootDir>/js/src/tests/dependencies/woocommerce/settings',
```
but still. 
6. What APIs should be mocked there, and how? This is again pushing the responsibility for extension developers to keep track of and maintain the duplicate of settings package API. (Keeping track of unpublished undocumented API, watching every implementation change to `woocommerce/woocommerce` repo - which is impossible work)

------

If we try to address those problems individually, I could think of a number of (mutually inclusive or exclusive) improvements:
1. Publish woocommerce jest config that provides sufficient [configs](https://github.com/woocommerce/woocommerce/blob/1550806efc513643d7a4ae7fed0bd723e2c62929/packages/js/internal-js-tests/src/setup-globals.js) and [mocks](https://github.com/woocommerce/woocommerce/blob/1550806efc513643d7a4ae7fed0bd723e2c62929/packages/js/internal-js-tests/src/mocks/woocommerce-settings.js), to run unit tests for an extension using settings package.
2. Change woocommerce/eslint to allow importing from not installed packages
3. Mention in woo/DEWP the origin of `@woocommerce/settings` and provide a link to its docs.
4. Remove what's currently under NPMs https://www.npmjs.com/package/@woocommerce/settings as I guess it's no longer maintained, relevant, or used, and put the actual settings package there. Even if it won't work standalone, without the entire WordPress, it's still better than today.
   As it will provide the API, it will provide intelisense, solve eslint errors, and possibly will work if used in WP environment. Or If I'd like to re-create another WooCommerce-like WP plugin, like experiment with WooCommerceLight without cloning the repo, but rather composing the modules we have.
5. Or if the settings package is desired to be non-modular, and absolutely not to be used without the rest of WooCommerce, then maybe publish the mock as `@woocommerce/settings` to solve only external use which is DEWP-related lining and testing.

**Comment by psealock** [comment]
Thats a nice summary (1-5) @tomalec and make sense to me. I ran into this issue while writing [this guide](https://developer.woocommerce.com/2020/02/20/extending-wc-admin-reports/), so it makes sense to address this issue especially as dev advocacy begins to create resources to guide external developers.

**Comment by samueljseay** [comment]
@psealock I personally have been thinking about this a bit. I think actually we should decommission this package altogether.

in `@woocommerce/data` we have the settings hooks which are great in a React context for getting the settings data. And in other contexts like plain JS or otherwise probably it makes sense just to access the global? (those contexts will be pretty rare anyway, they won't likely be someone using Vue or something, more like people using plain js or jquery).

**Comment by samueljseay** [comment]
@psealock bump on this, would be good to decide what to do with this task. If we can decommission the package its probably a quick fix we could bring into a sprint?

**Comment by psealock** [comment]
>I think actually we should decommission this package altogether.

We should decommission `@woocommerce/settings` npm package, is that what you are suggesting?

>And in other contexts like plain JS or otherwise probably it makes sense just to access the global?

I don't think plain JS will access this data, and if it does its a sign it needs to be updated to React.

Its not a real package, and yes I think we could remove it as well. Current usage is mostly `import { getAdminLink } from '@woocommerce/settings';` and all other instances of gathering data in "settings" is done through `@woocommerce/data`.

**Comment by samueljseay** [comment]
@psealock yes, I think the package can just be removed from npm and not released there since it doesn't represent the package in github.

We have one mention of it as an actual dependency, which could be migrated away I think https://github.com/woocommerce/woocommerce/blob/469432a1d3fdc6fbea9d1dcef631e9beb1b6ce56/packages/js/product-editor/package.json#L42

**Comment by tomalec** [comment]
What is the migration path for extensions that currently use `@woocommerce/settings` (like [GLA](https://github.com/woocommerce/google-listings-and-ads/search?q=%22%40woocommerce%2Fsettings%22&type=code))? Use `@woocommerce/data`:

```js
import { SETTINGS_STORE_NAME } from '@woocommerce/data';
import { useSelect } from '@wordpress/data';

// …
const settings = useSelect( select => {
	return select( SETTINGS_STORE_NAME ).getSettings();
```

or stick to `wcSettings` global?

**Comment by samueljseay** [comment]
@tomalec This is just my opinion so please feel free to push back or tell me why I'm wrong, but, yes I think migrating to `@woocommerce/data` makes the most sense personally.

What do you think? We can start by deprecating this package rather than removing it?

**Comment by tomalec** [comment]
I vote for deprecating to give us and other extension devs some time to migrate.

**Comment by samueljseay** [comment]
cc @psealock

**Comment by dkrebs** [comment]
Whats the state of this? I ran into the same issue. If you look at npmjs, the settings package is over 5 years old, but downloads were increasing at the end of 2023 until now. So it seems like many developers are confused as well.

The dev experience for external developers for wooCommerce Blocks is a bit frustrating :-(

**Comment by samueljseay** [comment]
@dkrebs Yes I agree it's confusing and I'm sorry to hear about your frustration.

The state is that package is deprecated but hasn't been marked as such.

@dkrebs  Have you got the [WooCommerce Dependency Extraction Webpack Plugin](https://github.com/woocommerce/woocommerce/tree/trunk/packages/js/dependency-extraction-webpack-plugin) setup in your environment? If you are having trouble accessing the globals you'll need to have that setup first. I realize it can be a bit daunting, if you need additional help you could reach out in the community slack too, it's easier to have a discussion there if you need extended technical help. https://join.slack.com/t/woocommercecommunity/shared_invite/zt-2kdclgfae-HpkiDuIOr_OvuWixcZWwvQ

**Comment by dkrebs** [comment]
@samueljseay I have the [WooCommerce Dependency Extraction Webpack Plugin](https://github.com/woocommerce/woocommerce/tree/trunk/packages/js/dependency-extraction-webpack-plugin) installed. The project it is building, that's fine. But there seems to be no way to get the types in my project as they are not exported by wooCommerce packages. Thus, there is no auto completion. It feels like developing in the 90s. Next, you need to build the whole wooCommerce project yourself to get source maps.

**Comment by samueljseay** [comment]
@dkrebs yep i agree. We have no consistent approach to publishing types right now unfortunately. Partially thats because in some areas of the project the migration to Typescript has been slow and done over time. 

I don't have a timeline for us to remedy this but I have raised it with the team.

**Comment by timmy5685** [comment]
Hi @tomalec - thanks for such a detailed report, it's really helpful.

I know this has been open for a very long time, and I appreciate your patience. We're currently going through some older WooCommerce issues, identifying what still needs fixing, and tidying up what's already fixed or out of scope.

I'd like to get an engineer's eyes on this before we make a call, so I've flagged it for developer feedback. Someone will take a proper look at the repro steps and how it's behaving now.

**Comment by vladolaru** [comment]
I checked this against current trunk and the current npm metadata. This still looks valid to me.

DEWP still treats `@woocommerce/settings` as a WooCommerce external alias rather than a normal package contract: the README maps it to [`wc['wcSettings']` / `wc-settings`](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/dependency-extraction-webpack-plugin/README.md#L43-L52), the plugin source maps `settings` to [`[ 'wc', 'wcSettings' ]`](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/dependency-extraction-webpack-plugin/src/index.js#L20-L55), and the externalized package list still includes [`@woocommerce/settings`](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/dependency-extraction-webpack-plugin/assets/packages.js#L24-L30). The Blocks build has the same [`wcSettings` / `wc-settings` mapping](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/blocks/bin/webpack-helpers.js#L14-L34), while the monorepo TypeScript config resolves the import to the local Blocks settings implementation via [`assets/js/settings/shared`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/blocks/tsconfig.base.json#L115-L117).

That local runtime path still does not match what external developers get from npm. The implementation reads `window.wcSettings` and builds `allSettings` from that global ([settings init](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/blocks/assets/js/settings/shared/settings-init.ts#L6-L127)), then exports helpers such as [`getSetting`, `getSettingWithCoercion`, `getAdminLink`, and `getPaymentMethodData`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/blocks/assets/js/settings/shared/utils.ts#L19-L147). The npm package is still `1.0.0`, deprecated, points back to the old WooCommerce Admin package location, and does not expose `types`/`typings` metadata.

We also still have local mitigations rather than a clear public contract: eslint treats `@woocommerce/settings` as an external/core module ([config](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/eslint-plugin/configs/recommended.js#L63-L66)), the components README tells Jest consumers to mock it manually ([docs](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/components/README.md#L36-L64)), internal tests carry a mock ([mock](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/internal-js-tests/src/mocks/woocommerce-settings.js#L1-L32)), and `experimental-products-app` still depends on [`@woocommerce/settings@1.0.0`](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/experimental-products-app/package.json#L52-L58) with local typings.

I would keep this open until we make an explicit tooling/API contract decision:

- If deprecated: update DEWP and extension-facing docs to state that `@woocommerce/settings` is a legacy external alias for `wc.wcSettings`, document the migration path where one exists (`@woocommerce/data` for React/admin settings store reads where appropriate, explicit global/runtime access where that is the supported answer), and clean up stale real-package dependencies/local typings.
- If supported: publish or generate a package artifact that matches the `wc-settings` runtime API, including TypeScript declarations and an official Jest mock/setup for external extensions.
- If only supported as a test/type helper: publish and document that explicitly, while making clear that browser runtime still comes from `wc-settings`.

Issue #39345 overlaps because bundled `@woocommerce` packages hit the same missing `@woocommerce/settings` contract, but it is broader. I would only close this as a duplicate if #39345 explicitly absorbs the DEWP/docs/npm/types/test acceptance criteria for `@woocommerce/settings`; otherwise #35603 is the focused issue for that contract.

**Comment by ralucaStan** [comment]
The package `@woocommerce/settings` is deprecated. The [June 2025 developer advisory](https://developer.woocommerce.com/2025/06/10/developer-advisory-deprecation-notice-for-unsupported-woocommerce-npm-packages/) lists it among the deprecated npm packages, and npm now flags it as no longer supported. So we're in your first option; the DEWP alias to wc.wcSettings stays, the npm package goes away.

What's left is the confusing part of the original report is the docs part, which I fixed in https://github.com/woocommerce/woocommerce/pull/68615
